src/Channels/UnownedBufferChannel.cs

```diff
@@ -0,0 +1,352 @@
+
+                // Allocate a new segment to hold the buffer being written.
+                var segment = new BufferSegment(buffer);
+                segment.End = buffer.Data.Length;
+
+                if (_head == null)
+                {
+                    // Update the head to point to the head of the buffer.
+                    _head = segment;
+                }
+                else if (_tail != null)
+                {
+                    // Add this segment to the end of the chain
+                    _tail.Next = segment;
+                }
+
+                // Always update tail to the buffer's tail
+                _tail = segment;
+
+                // Trigger the continuation
+                Complete();
+
+                // Wait for another read to come (or for the end of Reading, which will also trigger this gate to open) in before returning
+                await _readWaiting;
+
+                // We need to preserve any buffers that haven't been consumed
+                var current = _head;
+                while (current != null)
+                {
+                    current.Buffer = current.Buffer.Preserve(0, current.Buffer.Data.Length);
+                    current = current.Next;
+                }
+
+                // Cancel this task if this write is cancelled
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+        }
+
+        // Called by the WRITER via WriteAsync to run the READER
+        private void Complete()
+        {
+            // Don't need to interlock here. We have one reader and one writer and the ReadingStarted/ReadWaiting gates
+            // ensure that the read is blocked while we write and vice-versa.
+            var awaitableState = _awaitableState;
+            _awaitableState = _awaitableIsCompleted;
+
+            if (!ReferenceEquals(awaitableState, _awaitableIsCompleted) &&
+                !ReferenceEquals(awaitableState, _awaitableIsNotCompleted))
+            {
+                awaitableState();
+            }
+        }
+
+        // Called by the READER
+        private ReadableBuffer Read()
+        {
+            if (_consuming)
+            {
+                throw new InvalidOperationException("Cannot Read until the previous read has been acknowledged by calling Advance");
+            }
+            _consuming = true;
+
+            return new ReadableBuffer(new ReadCursor(_head), new ReadCursor(_tail, _tail?.End ?? 0));
+        }
+
+        // Called by the READER
+        void IReadableChannel.Advance(ReadCursor consumed, ReadCursor examined)
+        {
+            BufferSegment returnStart = null;
+            BufferSegment returnEnd = null;
+
+            if (!consumed.IsDefault)
+            {
+                returnStart = _head;
+                returnEnd = consumed.Segment;
+                _head = consumed.Segment;
+                _head.Start = consumed.Index;
```

Comment:
I would set head/tail back to null if `consumed.IsEnd` kind of how I changed SocketInput in https://github.com/aspnet/KestrelHttpServer/pull/1120. This will prevent the last UnownedBuffer from being "preserved" (i.e. copied into a new array) when WriteAsync continues.